Story #910 Standard Interfaces - Port Item Master

jrbirch99 opened the pull request:
Moved over functionality needed for edi item master interface. 

I changed where the UDES menu option is stored in UDES 14, in UDES 11 it was stored in udes_stock which felt wrong? 

Some of the EDI code in UDES 14/edi_record had been abstracted into another function, but it meant that you could not have multiple products with the same ref_id, so I changed it back to what it originally was in UDES 11.

jrbirch99 committed:
[IMP] edi: Validate attachment filenames against the format defined in glob.

Story/910

Signed off by: Jack Birch <jack.birch@unipart.io>

jrbirch99 committed:
[IMP] edi: Allow multiple products on an record to be processed.

Story/910

Signed off by: Jack Birch <jack.birch@unipart.io>

armandcela commented on addons/edi/models/__init__.py:
```diff
@@ -23,3 +23,5 @@
 from . import edi_partner_record
 from . import edi_raw_document
 from . import edi_raw_record
+
```
Nit: No needed the empty line

armandcela commented on addons/edi/models/edi_document.py:
```diff
@@ -84,6 +84,11 @@ def _default_project_id(self):
     )
 
+    # Optionally enforce filename globs
+    enforce_filename = fields.Boolean(string="Enforce Document Filenames",
```
Nit: I think indentation is not correct, try formatting

jrbirch99 replied:
Yeah good spot, I ran black formatting for the whole file and there were quite a few changes

armandcela commented on addons/edi/models/edi_record.py:
```diff
@@ -292,7 +286,10 @@ def prepare(self, doc, vlist):
         with self.statistics() as stats:
             try:
-                self.create(list(self._add_doc_id(doc, vlist)))
+                for record_vals in vlist:
```
Why creating inside the loop, i think for performance is better to create similar as it was before.

jrbirch99 replied:
The original code doesn't work properly, there are errors when there are duplicate product lines in the edi record.

armandcela replied:
`total_list` = [loop inside vlsit and update doc_id]
`self.create(total_list)`
`count = len(total_list)`

armandcela replied:
total_list = [{**d,{"doc_id": doc.id}} for d in vlist]

jrbirch99 replied:
This worked nicely.

armandcela commented on addons/edi/models/ir_attachment.py:
```diff
@@ -0,0 +1,38 @@
+                    globs = recs.mapped("glob")
+                    for input in doc.input_ids:
+                        passed = False
+                        for glob in globs:
```
I think here or in the doc string are needed some comments to explain what is doing and why is breaking when finds the first match

jrbirch99 replied:
I think the doc string at the top explains what it does quite concisely, maybe this needs refactoring in the future? It is a quite a big mess of "if" statements.

armandcela commented on addons/edi/models/ir_attachment.py:
```diff
@@ -0,0 +1,38 @@
+                    globs = recs.mapped("glob")
+                    for input in doc.input_ids:
+                        passed = False
+                        for glob in globs:
```
passed = bool(any(fnmatch.fnmatch(input.datas_fname, glob) for glob in globs))

armandcela commented on addons/edi/models/edi_record.py:
```diff
@@ -280,16 +295,15 @@ def prepare(self, doc, vlist):
-                    record_vals['doc_id'] = doc.id
-                    self.create(record_vals)
-                    count += 1
+                values_to_write = [
```
It looks a little confusing the variable name values_to_write which is used in .create() :(
Maybe smth like vals can do the job

PeterC10 commented on addons/edi/models/ir_attachment.py:
```diff
@@ -0,0 +1,38 @@
+    _inherit = "ir.attachment"
+
+    @api.constrains("res_model")
+    def check_filename(self):
```
Q: Do we need this? Can't see it being called anywhere...

jrbirch99 replied:
This function is called when res_model field of the ir.attachment model changes - when a record is attached to a document. There is an enforce_filename field in the edi.document model which is set for each document type, I ported this over as it is set as True for the Item master document.

I think "enforce_filename" is a UDES field, I am now thinking it should be "u_enforce_filename"

PeterC10 replied:
Sorry I completely missed the `api.constrains` decorator. Ignore this comment.

Although FYI it shouldn't be `u_`, but `x_` in this case, as it is a customisation, but it's not being added by a UDES repo. :)

PeterC10 replied:
Sorry this whole comment thread was a disaster on my part. We are in the `odoo-edi` repo, so this is the repo that adds `edi.document`, and as it is a new model we don't need any field prefixes.

Sorry again! Resolving now...

PeterC10 commented on addons/edi/views/edi_menu_views.xml:
```diff
@@ -13,6 +13,10 @@
 	      sequence="20"/>
 
+    <!-- Records/UDES menu -->
+    <menuitem id="record_udes_menu" name="UDES"
```
Q: Why are we adding a menu called `UDES` in `edi`? This module is used by UDES, but Odoo does not become UDES by installing this.

PeterC10 replied:
As discussed, let's move this to the module where we'll add UDES EDI records. :)

jrbirch99 committed:
[FIX] edi: Fix issue with elide relating to NewId.

Two identical values may be evaluated as different when part of them (while being the same value) needs to be constructed as another new record. Example: partner title record has not yet been constructed and the same partner is attempted to be created twice, this results in the partner being identified as different causing an attempt at creating the parnter twice, the unique constraint for a partner then causes the creation of a partner to be invalidated.

Story/910

Signed off by: Jack Birch <jack.birch@unipart.io>
